src/protocol/sort/apply.rs

```diff
@@ -0,0 +1,103 @@
+use embed_doc_image::embed_doc_image;
+/// Permutation reorders (1, 2, . . . , m) into (σ(1), σ(2), . . . , σ(m)).
+/// For example, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is reordered into (C, D, B, A) by σ.
+#[embed_doc_image("apply", "images/sort/apply.png")]
+#[allow(dead_code)]
+pub fn apply<T: Copy + Default>(source_indices: &[usize], values: &[T]) -> Vec<T> {
+    let mut applied_vec: Vec<T> = (0..source_indices.len())
+        .map(|_| Default::default())
+        .collect();
+    (0..source_indices.len()).for_each(|i| {
+        let idx = source_indices[i];
+        applied_vec[i] = values[idx];
+    });
+    applied_vec
+}
+
+/// To compute `apply_inv` on values, destination_indices(i) can be regarded as the destination of i, i.e., the i-th item
+/// is moved by `apply_inv` to be the σ(i)-th item. Therefore, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is
+/// reordered into (D, C, A, B).
+#[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+#[allow(clippy::module_name_repetitions, dead_code)]
+pub fn apply_inv<T: Copy + Default>(destination_indices: &[usize], values: &[T]) -> Vec<T> {
+    let mut applied_vec: Vec<T> = (0..destination_indices.len())
+        .map(|_| Default::default())
+        .collect();
+    (0..destination_indices.len()).for_each(|i| {
+        let idx = destination_indices[i];
+        applied_vec[idx] = values[i];
+    });
+    applied_vec
+}
+
+/// Applying inverse using heapsort which applies destination in-place and in a fixed time
```

Comment:
Can we similarly link out to some kind of documentation that explains the algorithm going on here?

Reply:
Removing this for now, will improve this based on Ryo's response